Add support for ACME endpoint(s)

zmousm opened the pull request:
* Add `cert_manager.client.Client.get_paginated()`, used for listing (searching) where Sectigo API doc implies use of pagination
* Let `Client.get()` also accept optional query parameters, so it can be reused with the new method
* Implement support for ACME account endpoint
* Add tests for ACME, `Client.get()` with parameters and `Client.get_paginated()`

coreone commented on cert_manager/client.py:
```diff
@@ -143,6 +143,32 @@ def get(self, url, headers=None):
+        while position == 0 or len(result_json) == page_size:
+            next_url = prepare_url(
+                url, {'position': position, 'size': page_size})
+            result = self.__session.get(next_url, headers=headers)
```
Since we are using sessions already, it would probably make more sense to use the built-in function `Session.prepare_request`.

```python
req = Request('GET', url)
while position == 0 or len(result_json) == page_size:
    next_req = self.__session.prepare_request(req)
    next_req.params = {'position': position, 'size': page_size}
    result = self.__session.send(next_req)
```

I haven't tested the above code, but the documentation seems to suggest this is the way to go:

https://requests.readthedocs.io/en/master/user/advanced/#prepared-requests

zmousm replied:
The `PreparedRequest` instantiated in your example does not expose `params` as a writable property, so in order to set query parameters one needs to use `prepare_url()` or the full `prepare()` method.

zmousm commented on cert_manager/_helpers.py:
```diff
@@ -10,7 +10,12 @@
 from requests.exceptions import HTTPError
+from requests import PreparedRequest
 
+def prepare_url(url, params={}):
```
Alright. Looking back the utility function is not necessary after all as there is indeed a more straightforward approach.

I think the docstring you pointed to cautions against _sending_ a manually instantiated `PreparedRequest` (I only used it to get the URL plus query string) rather than implying it is a private API -- both `Request.prepare()` and `Session.prepare_request()` work similarly so it seems safe.

coreone commented:
@zmousm I think this looks pretty good now.  If you pull the latest `master` branch and rebase your branch, it should clear up most of the linting errors.  The errors left related to your changes are as follows (in case you can't see the CircleCI output):

```
tests/test_acme.py:22: [R] R1725 Consider using Python 3 style super() without arguments [pylint]
tests/test_acme.py:110: [C] C0103 Variable name "qs" doesn't conform to snake_case naming style [pylint]
tests/test_acme.py:140: [C] E128 continuation line under-indented for visual indent [pycodestyle]
tests/test_acme.py:158: [C] E128 continuation line under-indented for visual indent [pycodestyle]
tests/test_acme.py:172: [C] E128 continuation line under-indented for visual indent [pycodestyle]
tests/test_acme.py:190: [C] E128 continuation line under-indented for visual indent [pycodestyle]
tests/test_acme.py:218: [C] E306 expected 1 blank line before a nested definition, found 0 [pycodestyle]
tests/test_acme.py:231: [C] E128 continuation line under-indented for visual indent [pycodestyle]
tests/test_acme.py:248: [C] E302 expected 2 blank lines, found 1 [pycodestyle]
cert_manager/acme.py:30: [R] R1725 Consider using Python 3 style super() without arguments [pylint]
cert_manager/acme.py:50: [R] R0913 Too many arguments (6/5) [pylint]
cert_manager/acme.py:50: [W] W0613 Unused argument 'org_id' [pylint]
cert_manager/acme.py:50: [W] W0613 Unused argument 'name' [pylint]
cert_manager/acme.py:50: [W] W0613 Unused argument 'acme_server' [pylint]
cert_manager/acme.py:50: [W] W0613 Unused argument 'cert_validation_type' [pylint]
cert_manager/acme.py:50: [W] W0613 Unused argument 'status' [pylint]
cert_manager/acme.py:86: [W] W0102 Dangerous default value {} as argument [pylint]
cert_manager/acme.py:126: [W] W0511 XXX: revisit [pylint]
cert_manager/acme.py:139: [W] W0511 XXX: revisit [pylint]
```

zmousm commented:
> @zmousm I think this looks pretty good now. If you pull the latest `master` branch and rebase your branch, it should clear up most of the linting errors. The errors left related to your changes are as follows (in case you can't see the CircleCI output):

Thanks, I can not see CircleCI indeed.

I believe I addressed most of these issues.

> ```
> cert_manager/acme.py:50: [R] R0913 Too many arguments (6/5) [pylint]
> cert_manager/acme.py:50: [W] W0613 Unused argument 'org_id' [pylint]
> cert_manager/acme.py:50: [W] W0613 Unused argument 'name' [pylint]
> cert_manager/acme.py:50: [W] W0613 Unused argument 'acme_server' [pylint]
> cert_manager/acme.py:50: [W] W0613 Unused argument 'cert_validation_type' [pylint]
> cert_manager/acme.py:50: [W] W0613 Unused argument 'status' [pylint]
> ```

I didn't switch to `**kwargs` previously because I was not sure how keyword arguments should be documented in the particular docstring style (it looks like reST/Sphinx, but it is not exactly so). Anyway.

> ```
> cert_manager/acme.py:126: [W] W0511 XXX: revisit [pylint]
> cert_manager/acme.py:139: [W] W0511 XXX: revisit [pylint]
> ```

I am still undecided about these and I would appreciate your feedback whether these two methods should return a boolean value.

zmousm commented:
I still have to complete the ACME tests. I need some time for that.

coreone commented:
CircleCI is looking good.  In other methods with return values that aren't 100% needed, I've often just return `result.json()` to return back whatever the API sends back just in case whomever called the method wanted to find out exactly what the API response was.  Thanks for the effort on this.  No rush on this.  Just wanted to alert you to the rebase/CircleCI bits.  I hope to port this to Actions soon to make all this more visible.

zmousm commented:
> CircleCI is looking good.

The `preflight-checks` seems to be failing again, but I can't see/say why.

> In other methods with return values that aren't 100% needed, I've often just return `result.json()` to return back whatever the API sends back just in case whomever called the method wanted to find out exactly what the API response was.

In this case there is no response body, the only meaningful thing returned from the API (for such PUT/DELETE requests) is the http code. So to return some data I would have to synthesize it... Your thoughts?

> No rush on this. Just wanted to alert you to the rebase/CircleCI bits. I hope to port this to Actions soon to make all this more visible.

Thanks, OK.

coreone commented:
The preflight-checks are just the two methods with the `#XXX` comments, so once those are gone, it should be good.  Since there is no data returned, boolean sounds great.  Thanks!

coreone commented:
As a separate note, could you change the base branch for this PR to `main`?

https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request

That is going to be our new default branch going forward.  Thanks!

coreone commented:
I migrated over to GitHub Actions, so if you rebase you should be able to see all the linting and unit tests through the GitHub UI now.

zmousm commented:
Tests are still WIP. I will finally squash the extra commits.

zmousm commented:
Tests are complete. As noted previously I intend to squash (down to 3 commits) after your review.

zmousm commented:
On a second thought some tests may also be due for client `get()` parameters and `get_paginated()`...

zmousm commented:
All done. Let me know if/when I should squash recent commits.

coreone commented:
Looks good to me.  Squash it down and I'll approve/merge.
